File: Apuntes/R/RegresionPolinomica.R

# Regresión polinomica

dataset = read.csv("C:/Users/Aaronga/Documents/GitHub/MachineLearningAZ/datasets/Part 2 - Regression/Section 6 - Polynomial Regression/Position_Salaries.csv")
dataset = dataset[, 2:3]

## Tratamiento de los NaN
"""
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x)mean(x,na.rm = T)),dataset$Age)
dataset$Salary = ifelse(is.na(dataset$Salary),
                        ave(dataset$Salary, FUN = function(x)mean(x,na.rm = T)),dataset$Salary)
"""

# codificar variables categoricas

"""
dataset$Country = factor(dataset$Country,
                         levels = c("France", "Spain", "Germany"),
                         labels = c(1,2,3))
dataset$Purchased = factor(dataset$Purchased,
                           levels = c("No","Yes"),
                           labels = c(0,1))
"""

# Dividir los datos en conjunto de entrenamiento y conjunto de test
"""
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
split
trainingSet = subset(dataset, split == T)
testingSet = subset(dataset, split == F)
"""

# Escalado de valores 

trainingSet[,2:3] = scale(trainingSet[,2:3])
testingSet[,2:3] = scale(testingSet[,2:3])

# Ajustar modelo lineal 

linReg = lm(Salary ~.,data = dataset)
summary(linReg)
# Ajustar modelo polinomico
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
dataset$Level5 = dataset$Level^5

polyReg = lm(Salary ~ ., data = dataset)
summary(polyReg)


# Prediccion de nuevos resultados con regresion lineal
y_pred = (predict(linReg, newdata = data.frame(Level = 6.5)))

# Prediccion de nuevos resultados con regresion polinomica
y_predPoly = (predict(polyReg, newdata = data.frame(Level = 6.5,
                                                Level2 = 6.5^2,
                                                Level3 = 6.5^3,
                                                Level4 = 6.5^4,
                                                Level5 = 6.5^5)))

# Visualizacion del modelo lineal 
library(ggplot2)
ggplot() + 
        geom_point(aes(x = dataset$Level, 
                       y = dataset$Salary), colour =  "red") +
        geom_line(aes(x=dataset$Level, 
                      y = predict(linReg, newdata = dataset)), colour = "blue") +
        ggtitle("Predicción lineal del sueldo en funcion del empleado") +
        xlab("Nivel del empleado")+
        ylab("Sueldo en USD")

# Visualización del modelo polinomico
x_grid =  seq(min(dataset$Level), max(dataset$Level),0.1)
ggplot() + 
        geom_point(aes(x = dataset$Level, 
                       y = dataset$Salary), colour =  "red") +
        geom_line(aes(x=x_grid, 
                      y = predict(polyReg, newdata = data.frame(Level = x_grid,
                                                                Level2 = x_grid^2,
                                                                Level3 = x_grid^3,
                                                                Level4 = x_grid^4,
                                                                Level5 = x_grid^5))), colour = "blue") +
        ggtitle("Predicción polinomica del sueldo en funcion del empleado") +
        xlab("Nivel del empleado")+
        ylab("Sueldo en USD")

File: Apuntes/Python/Regression (Cuantitativos)/RegresionPolinomica.py
# -*- coding: utf-8 -*-
"""
Created on Tue Oct 27 18:30:41 2020

@author: Aaronga
"""

# Regresión polinómica

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv("C:/Users/Aaronga/Documents/GitHub/MachineLearningAZ/datasets/Part 2 - Regression/Section 6 - Polynomial Regression/Position_Salaries.csv")
X = dataset.iloc[:,1:2].values # Para convertirlo a vector se usa [:,1:2]
y = dataset.iloc[:, 2].values
"""
# Tratamiento de los NaN 
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values="NaN", strategy="mean", axis = 0)
imputer = imputer.fit(X[:, 1:3])
X[:, 1:3]= imputer.transform(X[:,1:3])
print(X)

# Codificar datos categoricos
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelEncoderX = LabelEncoder()
X[:,0]=labelEncoderX.fit_transform(X[:,0])
onehotencoder = OneHotEncoder(categorical_features=[0])
X = onehotencoder.fit_transform(X).toarray()
labelEncoderY = LabelEncoder()
y=labelEncoderY.fit_transform(y)"""

                    # VARIABLE DUMMY 
                        # 0 0 0/ 0 1 0 / 0 0 1
# Dicidir el data set en conjunto de entrenamiento y conjunto de testing
 """                       
from sklearn.model_selection import train_test_split
XTrain, XTest, YTrain, YTest = train_test_split(X, y, test_size = 0.2, random_state = 0)
"""
# Escalado de variables 
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
XTrain = sc_X.fit_transform(XTrain)
XTest = sc_X.transform(XTest)
"""

# Ajustar regresion lineal

from sklearn.linear_model import LinearRegression
linRegression = LinearRegression()
linRegression.fit(X, y)

# Ajustar regresion polynomial

from sklearn.preprocessing import PolynomialFeatures
polyReg = PolynomialFeatures(degree=5)
X_poly = polyReg.fit_transform(X)
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)

# Visualización de los datos del modelo lineal
plt.scatter(X,y, color= "red")
plt.plot(X, linRegression.predict(X), color="blue")
plt.title("Salario de acuerdo al nivel (Lineal)")
plt.xlabel("Nivel en la empresa")
plt.ylabel("Salario")
plt.show()

# Visualizacion de los datos del modelo polinomico

X_grid = np.arange(min(X), max(X+0.1), 0.1)
X_grid = X_grid.reshape(len(X_grid),1)
X_polyGrid = polyReg.fit_transform(X_grid)
plt.scatter(X,y, color= "red")
plt.scatter(6.5,lin_reg2.predict(polyReg.fit_transform([[6.5]])), color= "green" )
plt.plot(X_grid, lin_reg2.predict(X_polyGrid), color="blue") # Parametro original X_poly 
plt.title("Salario de acuerdo al nivel (Polinomica)")
plt.xlabel("Nivel en la empresa")
plt.ylabel("Salario")
plt.show()

# Predicción correspondiente

linRegression.predict([[6.5]])

lin_reg2.predict(polyReg.fit_transform([[6.5]]))